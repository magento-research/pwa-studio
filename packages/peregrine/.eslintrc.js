const config = {
    parser: 'babel-eslint',
    extends: ['@magento'],
    rules: {
        "no-undef": "off"
    }
};

module.exports = config;
