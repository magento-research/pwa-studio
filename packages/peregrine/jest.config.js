module.exports = {
    displayName: 'Peregrine',
    setupFiles: ['<rootDir>/scripts/shim.js', '<rootDir>/scripts/fetch-mock.js']
};
