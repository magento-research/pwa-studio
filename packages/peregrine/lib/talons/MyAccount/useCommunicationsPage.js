import { useCallback } from 'react';
import { useMutation } from '@apollo/react-hooks';
import { useUserContext } from '../../context/user';
import { useAwaitQuery } from '../../hooks/useAwaitQuery';

export const useCommunicationsPage = props => {
    const {
        mutations: { setNewsletterSubscriptionMutation },
        queries: { getCustomerQuery }
    } = props;

    const [{ currentUser, isSignedIn }, { getUserDetails }] = useUserContext();

    const initialValues = {
        isSubscribed:
            currentUser && currentUser.hasOwnProperty('is_subscribed')
                ? currentUser.is_subscribed
                : false
    };

    const [
        subscription,
        { error: subscriptionError, loading: isSubmitting }
    ] = useMutation(setNewsletterSubscriptionMutation);
    const fetchUserDetails = useAwaitQuery(getCustomerQuery);

    const handleSubmit = useCallback(
        async ({ isSubscribed }) => {
            if (isSubscribed === undefined) {
                isSubscribed = initialValues.isSubscribed;
            }
            await subscription({
                variables: { is_subscribed: isSubscribed }
            });

            // Ensure old stores are updated with any new data.
            getUserDetails({ fetchUserDetails });
        },
        [subscription, fetchUserDetails, getUserDetails, initialValues]
    );

    return {
        formErrors: [subscriptionError],
        isDisabled: isSubmitting,
        isSignedIn,
        handleSubmit,
        initialValues
    };
};
