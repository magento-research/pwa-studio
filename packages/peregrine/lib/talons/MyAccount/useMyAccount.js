import { useCallback } from 'react';

export const useMyAccount = props => {
    const { onSignOut } = props;

    const handleSignOut = useCallback(() => {
        onSignOut();
    }, [onSignOut]);

    return {
        handleSignOut
    };
};
