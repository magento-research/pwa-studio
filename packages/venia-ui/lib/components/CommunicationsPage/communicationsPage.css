.root {
    padding: 20px;
    font-size: 0.875rem;
}

.header {
    padding-bottom: 10px;
}

.title {
    margin-bottom: 25px;
    text-align: center;
    font-weight: 700;
}

.message {
    background-color: rgb(var(--venia-grey));
    border-radius: 4px;
    font-size: 0.875rem;
    line-height: 1.25rem;
    padding: 1rem;
}

.message:empty {
    display: none;
}

.actions {
    display: grid;
    justify-items: left;
    margin-top: 1rem;
}

.error {
    composes: message;
    color: rgb(var(--venia-error));
}

.form {
    margin-top: 15px;
}

.subscribe {
    margin-top: 25px;
}

.subscribe_label {
    font-weight: bold;
    margin-bottom: 8px;
}
