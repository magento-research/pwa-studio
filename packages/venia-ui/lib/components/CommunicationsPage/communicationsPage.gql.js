import gql from 'graphql-tag';

export const SET_NEWSLETTER_SUBSCRIPTION = gql`
    mutation updateCustomer($is_subscribed: Boolean!) {
        updateCustomer(input: { is_subscribed: $is_subscribed }) {
            customer {
                id
                is_subscribed
            }
        }
    }
`;

export const GET_CUSTOMER = gql`
    query GetCustomer {
        customer {
            id
            email
            firstname
            lastname
            is_subscribed
        }
    }
`;

export default {
    mutations: {
        setNewsletterSubscriptionMutation: SET_NEWSLETTER_SUBSCRIPTION
    },
    queries: {
        getCustomerQuery: GET_CUSTOMER
    }
};
