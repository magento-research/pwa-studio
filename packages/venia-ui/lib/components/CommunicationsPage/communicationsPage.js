import React from 'react';
import { Form } from 'informed';
import { Redirect } from '@magento/venia-drivers';
import { mergeClasses } from '../../classify';
import defaultClasses from './communicationsPage.css';
import { useNewsletterSubscription } from '@magento/peregrine/lib/talons/MyAccount/useNewsletterSubscription';
import GET_CUSTOMER_QUERY from '../../queries/getCustomer.graphql';
import SUBSCRIBE_MUTATION from '../../queries/userSubscription.graphql';
import Button from '../Button';
import Checkbox from '../Checkbox';

const CommunicationsPage = props => {
    const classes = mergeClasses(defaultClasses, props.classes);

    const talonProps = useNewsletterSubscription({
        customerQuery: GET_CUSTOMER_QUERY,
        subscribeMutation: SUBSCRIBE_MUTATION
    });

    const {
        is_subscribed,
        isSignedIn,
        toggleSubscription,
        handleSubmit,
        errors,
        isDisabled
    } = talonProps;

    // Map over any errors we get and display an appropriate error.
    const errorMessage = errors.length
        ? errors
              .map(({ message }) => message)
              .reduce((acc, msg) => msg + '\n' + acc, '')
        : null;

    if (!isSignedIn) {
        return <Redirect to="/" />;
    }

    return (
        <div className={classes.root}>
            <div className={classes.header}>
                <h1 className={classes.title}>Communications</h1>
                <p>
                    We'd like to stay in touch. Please check the boxes next to
                    the communications you'd like to receive.
                </p>
            </div>
            <Form className={classes.form} onSubmit={handleSubmit}>
                <div className={classes.subscribe}>
                    <h5 className={classes.subscribe_label}>
                        Venia E-Newsletter
                    </h5>
                    <Checkbox
                        field="is_subscribed_venia"
                        label="Stay on the cutting edge of fashion; subscribe to the monthly Venia Newsletter."
                    />
                </div>
                <div className={classes.subscribe}>
                    <h5 className={classes.subscribe_label}>
                        Weekly Promotions notices
                    </h5>
                    <Checkbox
                        field="is_subscribed"
                        label="Save money with exclusive offers and promotions direct to your inbox."
                        fieldState={{
                            value: is_subscribed
                        }}
                        onClick={toggleSubscription}
                    />
                </div>
                <div className={classes.error}>{errorMessage}</div>
                <div className={classes.actions}>
                    <Button disabled={isDisabled} type="submit" priority="high">
                        {isDisabled ? 'Saving' : 'Save Changes'}
                    </Button>
                </div>
            </Form>
        </div>
    );
};

export default CommunicationsPage;
