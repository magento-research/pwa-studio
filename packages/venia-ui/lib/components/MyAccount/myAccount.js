import React from 'react';
import {
    Archive as HistoryIcon,
    LogOut as SignOutIcon,
    Bell as BellIcon
} from 'react-feather';
import { func, shape, string } from 'prop-types';

import { mergeClasses } from '../../classify';
import AccountLink from './accountLink';
import defaultClasses from './myAccount.css';
import { useMyAccount } from '@magento/peregrine/lib/talons/MyAccount/useMyAccount';

const PURCHASE_HISTORY = 'Purchase History';
const SIGN_OUT = 'Sign Out';
const NEWSLETTER_SUBSCRIPTION = 'Newsletter Subscription';

const MyAccount = props => {
    const classes = mergeClasses(defaultClasses, props.classes);

    const talonProps = useMyAccount({
        onSignOut: props.onSignOut,
        onRedirectLocation: props.onRedirectLocation
    });

    const { handleSignOut, subtitle, title, handleLinkLocation } = talonProps;

    return (
        <div className={classes.root}>
            <div className={classes.user}>
                <h2 className={classes.title}>{title}</h2>
                <span className={classes.subtitle}>{subtitle}</span>
            </div>
            <div className={classes.actions}>
                <AccountLink>
                    <HistoryIcon size={18} />
                    {PURCHASE_HISTORY}
                </AccountLink>
                <AccountLink
                    onClick={() => handleLinkLocation('/newsletter-manager')}
                >
                    <BellIcon size={18} />
                    {NEWSLETTER_SUBSCRIPTION}
                </AccountLink>
                <AccountLink onClick={handleSignOut}>
                    <SignOutIcon size={18} />
                    {SIGN_OUT}
                </AccountLink>
            </div>
        </div>
    );
};

export default MyAccount;

MyAccount.propTypes = {
    classes: shape({
        actions: string,
        root: string,
        subtitle: string,
        title: string,
        user: string
    }),
    onSignOut: func.isRequired
};
