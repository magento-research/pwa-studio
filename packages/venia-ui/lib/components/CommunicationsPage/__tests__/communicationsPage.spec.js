import React from 'react';
import { createTestInstance } from '@magento/peregrine';
import { useCommunicationsPage } from '@magento/peregrine/lib/talons/MyAccount/useCommunicationsPage';

import CommunicationsPage from '../communicationsPage';

jest.mock('@magento/peregrine/lib/talons/MyAccount/useCommunicationsPage');
jest.mock('../../../classify');

const handleSubmit = jest.fn().mockName('handleSubmit');

const emptyFormProps = {
    formErrors: [],
    handleSubmit,
    initialValues: { isSubscribed: true },
    isDisabled: false,
    isSignedIn: true
};

jest.mock('@magento/peregrine', () => {
    const useToasts = jest.fn(() => [
        { toasts: new Map() },
        { addToast: jest.fn() }
    ]);

    return {
        ...jest.requireActual('@magento/peregrine'),
        useToasts
    };
});

jest.mock('../../Head', () => ({ Title: () => 'Communications' }));

test('renders empty form without data', () => {
    useCommunicationsPage.mockReturnValue(emptyFormProps);

    const tree = createTestInstance(<CommunicationsPage />);
    expect(tree.toJSON()).toMatchSnapshot();
});

test('renders form error', () => {
    useCommunicationsPage.mockReturnValueOnce({
        ...emptyFormProps,
        formErrors: [{ setNewsletterSubscriptionError: 'Form Error' }]
    });

    const tree = createTestInstance(<CommunicationsPage />);
    expect(tree.toJSON()).toMatchSnapshot();
});

describe('renders prefilled form with data', () => {
    const initialValues = {
        isSubscribed: false
    };

    test('with enabled buttons', () => {
        useCommunicationsPage.mockReturnValueOnce({
            formErrors: [],
            handleSubmit,
            initialValues,
            isDisabled: false,
            isSignedIn: true
        });

        const tree = createTestInstance(<CommunicationsPage />);
        expect(tree.toJSON()).toMatchSnapshot();
    });

    test('with disabled buttons', () => {
        useCommunicationsPage.mockReturnValueOnce({
            formErrors: [],
            handleSubmit,
            initialValues,
            isDisabled: true,
            isSignedIn: true
        });

        const tree = createTestInstance(<CommunicationsPage />);
        expect(tree.toJSON()).toMatchSnapshot();
    });
});
