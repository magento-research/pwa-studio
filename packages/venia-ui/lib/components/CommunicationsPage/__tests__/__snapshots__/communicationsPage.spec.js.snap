// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`redirects when not authenticated 1`] = `
<mock-Redirect
  to="/"
/>
`;

exports[`renders empty form without data 1`] = `
<div
  className="root"
>
  Communications
  <h1
    className="title"
  >
    <mock-FormattedMessage
      defaultMessage="Communications"
      id="communicationsPage.communicationsText"
    />
  </h1>
  <p>
    <mock-FormattedMessage
      defaultMessage="We'd like to stay in touch. Please check the boxes next to the communications you'd like to receive."
      id="communicationsPage.optInText"
    />
  </p>
  <form
    className="form"
    onKeyDown={[Function]}
    onReset={[Function]}
    onSubmit={[Function]}
  >
    <div
      className="root"
    >
      <label
        className="label"
        htmlFor="isSubscribed"
      >
        Venia E-Newsletter
      </label>
      <label
        className="root"
      >
        <input
          checked={true}
          className="input"
          id="6323a4ef-b9aa-42aa-ba13-0e2434f8c3c7"
          name="isSubscribed"
          onBlur={[Function]}
          onChange={[Function]}
          type="checkbox"
        />
        <span
          className="icon"
        >
          <svg
            fill="none"
            height="24"
            stroke="currentColor"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <polyline
              points="9 11 12 14 22 4"
            />
            <path
              d="M21 12v7a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2V5a2 2 0 0 1 2-2h11"
            />
          </svg>
        </span>
        <span
          className="label"
        >
          Stay on the cutting edge of fashion; subscribe to the monthly Venia Newsletter.
        </span>
      </label>
      <p
        className="root"
      />
    </div>
    <div
      className="buttonsContainer"
    >
      <button
        className="root_highPriority"
        disabled={false}
        type="submit"
      >
        <span
          className="content"
        >
          Save Changes
        </span>
      </button>
    </div>
  </form>
</div>
`;

exports[`renders form error 1`] = `
<div
  className="root"
>
  Communications
  <h1
    className="title"
  >
    <mock-FormattedMessage
      defaultMessage="Communications"
      id="communicationsPage.communicationsText"
    />
  </h1>
  <p>
    <mock-FormattedMessage
      defaultMessage="We'd like to stay in touch. Please check the boxes next to the communications you'd like to receive."
      id="communicationsPage.optInText"
    />
  </p>
  <form
    className="form"
    onKeyDown={[Function]}
    onReset={[Function]}
    onSubmit={[Function]}
  >
    <div
      className="root"
    >
      <label
        className="label"
        htmlFor="isSubscribed"
      >
        Venia E-Newsletter
      </label>
      <label
        className="root"
      >
        <input
          checked={true}
          className="input"
          id="6b6123c7-ecef-43f7-a7ad-7bdc845581f3"
          name="isSubscribed"
          onBlur={[Function]}
          onChange={[Function]}
          type="checkbox"
        />
        <span
          className="icon"
        >
          <svg
            fill="none"
            height="24"
            stroke="currentColor"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <polyline
              points="9 11 12 14 22 4"
            />
            <path
              d="M21 12v7a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2V5a2 2 0 0 1 2-2h11"
            />
          </svg>
        </span>
        <span
          className="label"
        >
          Stay on the cutting edge of fashion; subscribe to the monthly Venia Newsletter.
        </span>
      </label>
      <p
        className="root"
      />
    </div>
    <div
      className="buttonsContainer"
    >
      <button
        className="root_highPriority"
        disabled={false}
        type="submit"
      >
        <span
          className="content"
        >
          Save Changes
        </span>
      </button>
    </div>
  </form>
</div>
`;

exports[`renders prefilled form with data with disabled buttons 1`] = `
<div
  className="root"
>
  Communications
  <h1
    className="title"
  >
    <mock-FormattedMessage
      defaultMessage="Communications"
      id="communicationsPage.communicationsText"
    />
  </h1>
  <p>
    <mock-FormattedMessage
      defaultMessage="We'd like to stay in touch. Please check the boxes next to the communications you'd like to receive."
      id="communicationsPage.optInText"
    />
  </p>
  <form
    className="form"
    onKeyDown={[Function]}
    onReset={[Function]}
    onSubmit={[Function]}
  >
    <div
      className="root"
    >
      <label
        className="label"
        htmlFor="isSubscribed"
      >
        Venia E-Newsletter
      </label>
      <label
        className="root"
      >
        <input
          checked={false}
          className="input"
          id="f5078365-3aa3-4a40-a92c-5dea0c566f0f"
          name="isSubscribed"
          onBlur={[Function]}
          onChange={[Function]}
          type="checkbox"
        />
        <span
          className="icon"
        >
          <svg
            fill="none"
            height="24"
            stroke="currentColor"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <rect
              height="18"
              rx="2"
              ry="2"
              width="18"
              x="3"
              y="3"
            />
          </svg>
        </span>
        <span
          className="label"
        >
          Stay on the cutting edge of fashion; subscribe to the monthly Venia Newsletter.
        </span>
      </label>
      <p
        className="root"
      />
    </div>
    <div
      className="buttonsContainer"
    >
      <button
        className="root_highPriority"
        disabled={true}
        type="submit"
      >
        <span
          className="content"
        >
          Saving
        </span>
      </button>
    </div>
  </form>
</div>
`;

exports[`renders prefilled form with data with enabled buttons 1`] = `
<div
  className="root"
>
  Communications
  <h1
    className="title"
  >
    <mock-FormattedMessage
      defaultMessage="Communications"
      id="communicationsPage.communicationsText"
    />
  </h1>
  <p>
    <mock-FormattedMessage
      defaultMessage="We'd like to stay in touch. Please check the boxes next to the communications you'd like to receive."
      id="communicationsPage.optInText"
    />
  </p>
  <form
    className="form"
    onKeyDown={[Function]}
    onReset={[Function]}
    onSubmit={[Function]}
  >
    <div
      className="root"
    >
      <label
        className="label"
        htmlFor="isSubscribed"
      >
        Venia E-Newsletter
      </label>
      <label
        className="root"
      >
        <input
          checked={false}
          className="input"
          id="822aaa66-7c4c-4a6a-9100-83e2e3c64b89"
          name="isSubscribed"
          onBlur={[Function]}
          onChange={[Function]}
          type="checkbox"
        />
        <span
          className="icon"
        >
          <svg
            fill="none"
            height="24"
            stroke="currentColor"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <rect
              height="18"
              rx="2"
              ry="2"
              width="18"
              x="3"
              y="3"
            />
          </svg>
        </span>
        <span
          className="label"
        >
          Stay on the cutting edge of fashion; subscribe to the monthly Venia Newsletter.
        </span>
      </label>
      <p
        className="root"
      />
    </div>
    <div
      className="buttonsContainer"
    >
      <button
        className="root_highPriority"
        disabled={false}
        type="submit"
      >
        <span
          className="content"
        >
          Save Changes
        </span>
      </button>
    </div>
  </form>
</div>
`;
